1_dynamic/future.k

```diff
@@ -16,4 +18,465 @@ endmodule
+*/
+  rule <k> performCall($clinitClosure, _) => create(Class) </k>
+       <baseClass> Class:Id </baseClass>
+       <envStack> .List => ListItem((Object, <env>.Map</env>)) </envStack>
+
+/*
+  Method calls desugar into (blocking) future invocations
+*/
+
+  rule O:ObjectVal . Method:Id(Vs:Vals) 
+    => (O ! Method(Vs)).get
+  
+/*
+  A super call desugars into a normal call using a fake object
+  having the super class as a base.
+*/
+
+  rule <k> super . Method:Id(Vs:Vals)
+        => object(Id, Class', Class') . Method(Vs)
+       ...</k>
+      <crntClass> Class </crntClass>
+      <location> Id:Int </location>
+      <className> Class </className>
+      <extends> Class':Id </extends>
+
+/* Function calls desugar into future calls on the current object
+*/
+  rule <k>  methodClosure(Type,Class,OL,Ps,S) (Vs:Vals)
+        => future(F).get ...</k>
+       <location> OId </location>
+       (. => <future>...
+               <fid> F </fid> <oid> OId </oid>
+               <closure> methodClosure(Type,Class,OL,Ps,S) </closure>
+               <arguments> Vs </arguments>
+             ...</future>)
+       <nextFutureId> F:Int => F +Int 1 </nextFutureId>
+
+/*
+  The rules for blocking/unblocking tasks:
+*/
+
+  // A thread blocks while waiting for a future to complete
+  rule (<thread> <k> future(Id).get ~> K </k> T:Bag </thread> => . ) 
+       (. => <blocked-thread> <guard> future(Id).get ~> K </guard> 
+                              <blocked-thread-conf> T </blocked-thread-conf>
+             </blocked-thread>)
+       <fid> Id </fid>
+       <state> S:State </state>
+       requires S =/=K completed [futureBlock]
+
+  // A thread unblocks when the future is done
+  rule (. => <thread> <k> future(Id).get ~> K </k> T:Bag </thread>) 
+       (<blocked-thread> <guard> future(Id).get ~> K </guard> 
+                         <blocked-thread-conf> T </blocked-thread-conf>
+        </blocked-thread> => .)
+       <taskState> _ => active </taskState>
+       <fid> Id </fid>
+       <state> completed </state>
+
+/*
+  The rule that transforms an active task into an waiting task
+  * If all threads are blocked, we suspend the task.
+  * We use -1 for futureId to signal that there is no active task
+  * task-guards is a collection of all guards from the blocked threads
+*/
+  rule <futureId> F => -1 </futureId>
+       <blocked-threads> BT:Bag => . </blocked-threads>
+       <taskState> active </taskState>
+       <threads> . </threads>
+       (. => <waiting-task> <fId> F </fId>
+                            <task-guards> getGuards(BT) </task-guards> 
+                            <waiting-task-conf> BT </waiting-task-conf> 
+             </waiting-task>)
+      requires notBool(BT ==Bag .Bag)
+ 
+/*
+  The rule transforming a waiting task into an active one. 
+  * When any of the task guards is completed, we can awake a task
+    if there is no active task.
+*/
+  rule <blocked-threads> . => BT </blocked-threads>
```

Comment:
It seems to me that the commnet does not describe exactly what the rule does (was it changed?)